### PR TITLE
Deprecate `Trainer.gpus`

### PR DESCRIPTION
**Pull request by DuYicong515**
## What does this PR do?
Part of #12171: deprecate device related trainer properties: gpus
Trainer argument `gpus` is also being deprecated in #10410 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

File: tests/models/test_gpu.py [hunk 1]
```diff
@@ -190,10 +191,12 @@ def test_torchelastic_gpu_parsing(mocked_device_count, mocked_is_available, gpus
     sanitizing the gpus as only one of the GPUs is visible."""
     trainer = Trainer(gpus=gpus)
     assert isinstance(trainer._accelerator_connector.cluster_environment, TorchElasticEnvironment)
-    assert trainer.gpus == gpus
```
Comment (carmocca) [inline tests/models/test_gpu.py:193]: Why not keep the assert with `num_devices`?

File: tests/deprecated_api/test_remove_1-8.py [hunk 2]
```diff
@@ -1107,3 +1107,24 @@ def on_save_checkpoint(self, trainer, pl_module, checkpoint):
...
+)
+def test_trainer_gpus(monkeypatch, trainer_kwargs):
+    monkeypatch.setattr(torch.cuda, "is_available", lambda: True)
+    monkeypatch.setattr(torch.cuda, "device_count", lambda: 4)
+    trainer = Trainer(**trainer_kwargs)
+    with pytest.deprecated_call(
+        match="`Trainer.gpus` was deprecated in v1.6 and will be removed in v1.8."
+        " Please use `Trainer.num_devices` or `Trainer.device_ids` to get device information instead."
+    ):
+        assert trainer.gpus == trainer_kwargs.get("devices")
```
Comment (carmocca) [inline tests/deprecated_api/test_remove_1-8.py:1130]:
```suggestion
        assert trainer.gpus == trainer_kwargs["devices"]
```

**Review by carmocca — APPROVED** [review]

File: tests/models/test_gpu.py (hunk 1 above)
Comment (DuYicong515) [reply tests/models/test_gpu.py:193]: Could add that.

**Review by akihironitta — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
